test/src/specs/tx_pool/dao.rs

```diff
@@ -0,0 +1,403 @@
+use crate::utils::{assert_send_transaction_fail, is_committed};
+use crate::{Net, Node, Spec};
+use ckb_core::block::Block;
+use ckb_core::script::{Script, ScriptHashType};
+use ckb_core::transaction::{CellInput, CellOutput, OutPoint, Transaction, TransactionBuilder};
+use ckb_core::{BlockNumber, Bytes, Capacity};
+use ckb_resource::CODE_HASH_DAO;
+use ckb_test_chain_utils::always_success_cell;
+use numext_fixed_hash::H256;
+
+const SYSTEM_CELL_ALWAYS_SUCCESS_INDEX: u32 = 1;
+const SYSTEM_CELL_DAO_INDEX: u32 = 3;
+const WITHDRAW_WINDOW_LEFT: u64 = 10;
+const WITHDRAW_HEADER_INDEX: u8 = 0;
+
+pub struct DepositDAO;
+
+impl Spec for DepositDAO {
+    fn run(&self, net: Net) {
+        let node0 = &net.nodes[0];
+        node0.generate_blocks(2);
+
+        // Deposit DAO
+        {
+            let transaction = deposit_dao_transaction(node0);
+            ensure_committed(node0, &transaction);
+        }
+
+        // Deposit DAO without specifying `block_hash` within input
+        {
+            // deposit dao transaction without `block_hash`
+            let transaction = {
+                let out_point_without_block_hash = {
+                    let tx_hash = node0.get_tip_block().transactions()[0].hash().to_owned();
+                    OutPoint::new_cell(tx_hash, 0)
+                };
+                TransactionBuilder::from_transaction(deposit_dao_transaction(node0))
+                    .inputs_clear()
+                    .input(CellInput::new(out_point_without_block_hash, 0))
+                    .build()
+            };
+            ensure_committed(node0, &transaction);
+        }
+    }
+}
+
+pub struct WithdrawDAO;
+
+impl Spec for WithdrawDAO {
+    fn run(&self, net: Net) {
+        let node0 = &net.nodes[0];
+        node0.generate_blocks(2);
+
+        let deposited = {
+            let transaction = deposit_dao_transaction(node0);
+            ensure_committed(node0, &transaction)
+        };
+        let transaction = withdraw_dao_transaction(node0, deposited.clone());
+        ensure_committed(node0, &transaction);
+    }
+}
+
+pub struct DepositDAOTwice;
```

Comment:
I think it's withdrawing many times, is not deposit DAO twice, the name is confusing.

Reply:
OK. let's change it to `WithdrawAndDepositDAOWithinSameTx`.